### ISSUE TITLE
Tracking issue for collections reform part 2 (RFC 509)

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#19986]
[RFC 509](https://github.com/rust-lang/rfcs/pull/509)

**Comment by TimDumol** [comment]
I gave a shot at implementing `append()` and `split_off()` for DList over here: https://github.com/rust-lang/rust/pull/20406

**Comment by csouth3** [comment]
Looks like ExactSizeIterator is (most likely) getting decoupled from DoubleEndedIterator.  This presents the possible solution of just implementing ExactSize for the iterators in Hash{Map, Set} without DoubleEnded, and thus they gain `len()` and we avoid the DoubleEnded/`next_back()` issue that didn't make much sense earlier.  Just tossing that in as a possibility.

**Comment by brson** [comment]
Still needs `apppend` and `split_off`. Easy newbie bug.

**Comment by csouth3** [comment]
Slightly unrelated because it's not really a newbie bug, but looks like the RFC also says that Vec should be moved from `std::vec` to `std::collections::vec`...that'll be a fun one :p

**Comment by Gankro** [comment]
Did it? That change has been rejected any time I've brought it up.

**Comment by csouth3** [comment]
Oh, it did...but we can just _ignore_ that part XD.
It's under the Misc Stabilization part.

**Comment by tshepang** [comment]
@Gankro why the rejection? I found it odd that I could not find it in `std::collections`.

**Comment by Gankro** [comment]
@aturon can give a better explanation

**Comment by msiemens** [comment]
I'll look whether I can implement `append` and `split_off` for the map/set collections.

@Gankro: The todo is a little outdated: `append` and `split_off` have been implemented for `Vec` and `RingBuf`.

**Comment by Gankro** [comment]
@msiemens fixed, thanks!

**Comment by SDX2000** [comment]
I was told to leave a comment here instead of opening a new issue:
Rename string::push_str to string::concat #20857

Please consider renaming string::push_str() to string::concat(). Push_X() in the context of a collection means X will be added to the container as a single entity. Hence push_str() is meaningless for strings since a string can only contain characters (char != [char]).

If it is too late to rename then consider adding a method called concat() anyway.

Also consider renaming Vec::push_all() to Vec::concat() so that the name is consistent with String::concat().

**Comment by Gankro** [comment]
`push_all` isn't stabilized, and will likely (not definite yet) be replaced with `extend` per https://github.com/rust-lang/rfcs/pull/839 

Seems like the same could be done to push_str.

**Comment by aturon** [comment]
@Gankro 

> Seems like the same could be done to push_str.

Unfortunately the perf problem there is more fundamental: going through a character iterator involves UTF-8 decoding and re-encoding.

**Comment by Gankro** [comment]
Aha, fair enough.

Well, you could specifically  `impl Extend<&str>` modulo coherence.

**Comment by aturon** [comment]
> Well, you could specifically  `impl Extend<&str>` modulo coherence.

Sadly, I don't think that works either, because then you need to provide an *iterator* over string slices, which would make for poor ergonomics.

**Comment by Gankro** [comment]
Oh yeah no. Sorry long day I had forgotten how Extend is even implemented.

Extend specifically takes `O: IntoIterator<Item=T>`. With *specialization* you could optimize to do something special for `O = &str`, right?

**Comment by aturon** [comment]
> Extend specifically takes `O: IntoIterator<Item=T>`. With *specialization* you could optimize to do something special for `O = &str`, right? 

Yes, that's right.

**Comment by bfops** [comment]
A portion of the entry API changes were [reverted](https://github.com/rust-lang/rust/commit/25eada15740fbe12ee2cae7fc6fe8e2c228b699d). Isn't it still legitimate to only want to create a key on insert into a `VacantEntry`, rather than on every call to `entry`?

**Comment by aturon** [comment]
@bfops 

> A portion of the entry API changes were [reverted](https://github.com/rust-lang/rust/commit/25eada15740fbe12ee2cae7fc6fe8e2c228b699d). Isn't it still legitimate to only want to create a key on insert into a `VacantEntry`, rather than on every call to `entry`?

Absolutely. The previous change tried changing the API to favor this case, but it ended up having pretty negative consequences overall (in terms of forcing more allocations and ergonomic pain, IIRC).

The eventual plan is to use a trait like `Into<Cow<T>>` to seamlessly support both cases. Unfortunately, we can't take that step just yet, due to limitations in the trait system (coherence). Once we have specialization for trait implementations, this should become possible.

TL;DR: Yes, it's an important case, and we believe we can encompass it after 1.0.

**Comment by jooert** [comment]
@Gankro `append` and `split_off` for `VecMap` were implemented in #22494, could you update the todo list?

Oh, and the names of some collections have changed (https://github.com/rust-lang/rfcs/blob/master/text/0580-rename-collections.md#detailed-design), for example `Bitv` is now `BitVec`

**Comment by alexcrichton** [comment]
I believe this has all since been implemented, so closing.

**Comment by jooert** [comment]
Actually `append` and `split_off` have not been implemented for `BTreeMap` and `BTreeSet` yet.

**Comment by alexcrichton** [comment]
Ah good point.

**Comment by xosmig** [comment]
I started working on `append` and `split_off` methods for `BTreeMap` and `BTreeSet`.

**Comment by apasel422** [comment]
@xosmig #32466 already covers `BTreeMap::append`.

**Comment by xosmig** [comment]
@apasel422 OK. Then only `append`.

**Comment by xosmig** [comment]
Hello! I've faced some problem.

**Comment by xosmig** [comment]
Hello!
I've faced with problems.
To recalc sizes of trees after `split_off` using O(log n) operations, it's necessary to know exact size of each subtree.
So, there are some different solutions:

1) add `size` field to the `LeafNode` structure.

I think, benefits are great.

First, it lets us to do very flexible splits and merges with only O(log n) operations.
Now it's possible to merge trees using O(log n) operations, but it it not so interesting without fast split.

Second, some other interesting and usefull functions need it.
For example: `kth_element` or "get the number of elements between two values". Both with O(log n) operations.

Third, as I understand (but i am not totally sure), it lets us to implement the `rope` ( https://en.wikipedia.org/wiki/Rope_(data_structure) ) in easy way. Just as BTreeMap<(), V>.

But overhead is notable (4 or 8 bytes for each node, depending on machine word size). I think, it is possible to embed this sizes without memory overhead for 64-bit machines (due to aligning).

2) Implement `split_off` with O(log n + "size of the one of the results")

2.1) Split with O(log n) operations and then calculate size of the one of the results.

2.2) There is another algorithm which doesn't use the fast algorithm with the same complexity, but (as I understand) with greater constant.

3) Implement `split_off` with O(log n + "size of the smallest one")

Split using O(log n) operations and then calculate sizes of the trees via 2 alternating iterators, but stop when one of the sizes is counted.

**Comment by xosmig** [comment]
@apasel422, what do you think?

**Comment by apasel422** [comment]
@xosmig This isn't my area of expertise, but I'd recommend opening a discussion thread on https://internals.rust-lang.org/ or submitting a pull request for whichever option is easiest and CCing @jooert and @gereeter. You might also want to take a look at #26227.

**Comment by xosmig** [comment]
@apasel422 The main question is "Is this overhead okey or not?". So, who can decide it?

**Comment by gereeter** [comment]
First, note that option one can be slightly optimized by only storing composite sized in internal nodes.

I'm not a big fan of option one - I expect `split_off` to be a fairly rare operation, and slowing down everything else and increasing the memory footprint solely for its benefit does not seem like a good idea. Since we explicitly support large collections (as large as can fit in memory, basically), the size field would have to be a `usize`, which cannot be combined with other fields. The application to ropes is interesting, but really wants a separate implementation - the `BTreeMap` interface is totally wrong for the job (though the node module might be useful). `kth_element` and measuring the length of a range seems useful, but again, I see them as niche use cases that really want a specialized sequence data structure.

I don't have strong feelings between options two and three. I think that a variation on option two, where the choice of which side to count is made based on which edge was descended at the root node, would probably be the best. It doesn't have the overhead of option three, but still wont take long if you split near the edges.

There is also a fourth option, namely dropping support for constant time calls to `len()`. This is a somewhat more radical change, but it would easily allow for `split_off` in `O(log n)` and avoid the question of how to calculate size completely. Moreover, while it might inconvenience users, it is easy for users to wrap a BTreeMap with something that does track length, recovering the behavior for anyone who actually needs accurate lengths. I suspect that there is much higher demand for `is_empty` than there is for `len`.

**Comment by xosmig** [comment]
@gereeter 
About `rope`: it can be implemented as `BTree` with implicit key. It needs just 2 more usefull methods for BTreeMap: `split_by_number` (splits off the first `number` of ements) and `erase_by_number`.

**Comment by xosmig** [comment]
@gereeter 
"First, note that option one can be slightly optimized by only storing composite sized in internal nodes."
Yes, that's great! There are about (a rough estimate, in fact, on average, more) `6 / 7` leafs. So only `1 / 7` of all nodes should contain 1 extra usize. So, only about 1 / 63 n extra usizes. It's nothing, isn't it?
Now the size of `InternalNode` is about `11 * (size_of::<V>() + size_of::<K>()) + 112`. So, 8 extra bytes would be okay I think.

**Comment by xosmig** [comment]
@gereeter 
If I didn't make any mistake, I am totally assured that each Internal node must contain itself size.

**Comment by xosmig** [comment]
If I didn't make any blunder, I am assured that each Internal node should contain itselfs subtree size.

**Comment by xosmig** [comment]
There are some average values above. The worst case is: `(1 / 5)n` nodes and `1 / 6` internal nodes. So `(1 / 30)n` extra usizes, about 2 extra bits per key on a 64-bit architecture and about 1 extra bit per key on a 32-bit architecture.
